modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+  }
+
+  /**
+   * Returns an immutable set containing the currencies held within this matrix.
+   *
+   * @return the currencies in this matrix
+   */
+  public ImmutableSet<Currency> getCurrencies() {
+    return currencies.keySet();
+  }
+
+  @Override
+  public String toString() {
+    return getCurrencies() + " - " + Stream.of(rates).map(Arrays::toString).collect(Collectors.joining());
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * currencies.hashCode() + Arrays.deepHashCode(rates);
+  }
+
+  @Override
+  public boolean equals(final Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null || getClass() != obj.getClass()) {
+      return false;
+    }
+    FxMatrix other = (FxMatrix) obj;
+    return currencies.equals(other.currencies) && Arrays.deepEquals(rates, other.rates);
+  }
+
+  /**
+   * Creates a {@link Collector} that allows a Map of currency pair -> rates
+   * to be streamed and collected into a new {@code FxMatrix}.
+   *
+   * @return a collector for creating an {@code FxMatrix} from a stream
+   */
+  public static Collector<? super Pair<CurrencyPair, Double>, Builder, FxMatrix> pairsToFxMatrix() {
+    return collector((builder, pair) ->
+        builder.addRate(pair.getFirst(), pair.getSecond()));
+  }
+
+  /**
+   * Creates a {@link Collector} that allows a collection of pairs each containing
+   * a currency pair and a rate to be streamed and collected into a new {@code FxMatrix}.
+   *
+   * @return a collector for creating an {@code FxMatrix} from a stream
+   */
+  public static Collector<? super Map.Entry<CurrencyPair, Double>, Builder, FxMatrix> entriesToFxMatrix() {
+    return collector((builder, entry) ->
+        builder.addRate(entry.getKey(), entry.getValue()));
+  }
+
+  private static <T> Collector<T, Builder, FxMatrix> collector(BiConsumer<Builder, T> accumulator) {
+    return Collector.of(
+        FxMatrix::builder,
+        accumulator,
+        Builder::merge,
+        Builder::build);
+  }
+
+  /**
+   * Builder class for FxMatrix. Can be created either by the static
+   * {@link FxMatrix#builder()} or from an existing {@code FxMatrix}
+   * instance by calling {@link FxMatrix#toBuilder()}.
+   */
+  public static final class Builder {
+
+    /**
+     * The minimum size of the FX rate matrix. This is intended such
+     * that a number rates of can be added without needing to resize.
+     */
+    private static final int MINIMAL_MATRIX_SIZE = 8;
+
+    /**
+     * The currencies held by the builder pointing to their position
+     * in the rates array. An ordered map is used so that it retains
+     * order which means the {@code toString} method of {@code FxMatrix}
+     * is clearer.
+     */
+    private final LinkedHashMap<Currency, Integer> currencies;
+
+    /**
+     * A 2 dimensional array holding the rates. Each row of the array holds the
+     * value of 1 unit of Currency (that the row represents) in each of the
```

Comment:
Is the most common use case for this `double[][]` aligned with access pattern? I.e. Is data accessed row wise more often than column?